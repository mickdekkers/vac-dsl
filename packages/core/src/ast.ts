// TODO: allow optional meta field on all ASTNodes
export interface ASTNode {
  type: string
  loc: {
    start: {
      index: number
    }
  }
}

export interface Literal extends ASTNode {
  type: 'Literal'
  value: string
}

export interface Identifier extends ASTNode {
  type: 'Identifier'
  name: string
}

export interface VariableDeclaration extends ASTNode {
  type: 'VariableDeclaration'
  id: Identifier
  value: Literal
}

export type Node = Identifier | Literal

export interface NodeList extends ASTNode {
  type: 'NodeList'
  nodes: Node[]
}

export interface Property extends ASTNode {
  type: 'Property'
  name: string
  value: string | number
}

export interface PropertyList extends ASTNode {
  type: 'PropertyList'
  properties: Property[]
}

export interface EdgeChain extends ASTNode {
  type: 'EdgeChain'
  nodeLists: NodeList[]
  properties: PropertyList | null
}

export interface Comment extends ASTNode {
  type: 'Comment'
  value: string
  raw: string
}

export type Statement = VariableDeclaration | EdgeChain | Comment

export interface AST extends ASTNode {
  type: 'Program'
  meta: {
    [key: string]: any
  }
  body: Statement[]
}
