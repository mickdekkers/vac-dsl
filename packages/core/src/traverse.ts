import { orderedProps, aliases } from './ordered-ast-props'
import { ASTNode } from './ast'

export type VisitorFunction = (node: ASTNode, parent: ASTNode | null) => void

export type NodeVisitor =
  | VisitorFunction
  | {
      enter?: VisitorFunction
      leave?: VisitorFunction
    }

export interface Visitor {
  [key: string]: NodeVisitor
}

const getNodeVisitors = (nodeType: string, visitor: Visitor): NodeVisitor[] => {
  let nodeVisitors: NodeVisitor[] = []

  if (visitor[nodeType]) {
    nodeVisitors.push(visitor[nodeType])
  }

  // Check Visitor for aliases of the nodeType
  const nodeAliases = aliases.get(nodeType) || []

  // console.log(typeof nodeAliases.forEach, nodeAliases)
  nodeAliases.forEach(alias => {
    if (visitor[alias]) {
      nodeVisitors.push(visitor[alias])
    }
  })

  // Check Visitor for catch-all
  if (visitor['*']) {
    nodeVisitors.push(visitor['*'])
  }

  return nodeVisitors
}

export const traverse = (root: ASTNode, visitor: Visitor) => {
  // TODO: validate Visitor

  const visit = (node: ASTNode, parent) => {
    // Base case
    if (!node || typeof node.type !== 'string') {
      return
    }

    // Get visitors that are interested in current node type
    const nodeVisitors = getNodeVisitors(node.type, visitor)

    // Trigger enter visitors
    nodeVisitors.forEach(nodeVisitor => {
      if (typeof nodeVisitor === 'function') {
        nodeVisitor(node, parent)
      } else if (typeof nodeVisitor.enter === 'function') {
        nodeVisitor.enter(node, parent)
      }
    })

    // Get list of sub-nodes to recurse into
    const nodeProps = orderedProps.get(node.type)
    if (nodeProps === undefined) {
      throw new Error(`Unknown node type "${node.type}"`)
    }

    // Recurse
    nodeProps.forEach(nodeProp => {
      const child: ASTNode[] | ASTNode = node[nodeProp]
      if (Array.isArray(child)) {
        child.forEach(entry => visit(entry, node))
      } else {
        visit(child, node)
      }
    })

    // Trigger leave visitors
    nodeVisitors.forEach(nodeVisitor => {
      if (
        typeof nodeVisitor !== 'function' &&
        typeof nodeVisitor.leave === 'function'
      ) {
        nodeVisitor.leave(node, parent)
      }
    })
  }

  visit(root, null)
}
