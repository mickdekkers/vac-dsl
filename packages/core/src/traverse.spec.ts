import { AST } from './ast'
import { traverse } from './traverse'
import { ast as basicAst } from './__tests__/fixtures/programs/basic-literals'
import { ast as threeTwoThreeAst } from './__tests__/fixtures/programs/3-to-2-to-3-variables'
import { ast as propertiesMultiAst } from './__tests__/fixtures/programs/properties-multi'
import * as _ from 'lodash'

describe('traverse', () => {
  it('visits all nodes of type specified in visitor', () => {
    {
      const visitedNodes: string[] = []
      traverse(basicAst, {
        Identifier: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes.length).toBe(0)
    }

    {
      const visitedNodes: string[] = []
      traverse(threeTwoThreeAst, {
        Identifier: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes.every(node => node === 'Identifier')).toBe(true)
      expect(visitedNodes.length).toBe(16)
    }

    {
      const visitedNodes: string[] = []
      traverse(propertiesMultiAst, {
        Identifier: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes.every(node => node === 'Identifier')).toBe(true)
      expect(visitedNodes.length).toBe(7)
    }
  })

  it('accepts aliases', () => {
    {
      const visitedNodes: string[] = []
      traverse(basicAst, {
        Statement: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }

    {
      const visitedNodes: string[] = []
      traverse(threeTwoThreeAst, {
        Statement: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }

    {
      const visitedNodes: string[] = []
      traverse(propertiesMultiAst, {
        Statement: node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }
  })

  it('visits all nodes in order when using * catch-all', () => {
    {
      let visitedLog = ''
      let indent = 0
      traverse(basicAst, {
        '*': {
          enter: (node, parent) => {
            visitedLog += ' '.repeat(indent) + node.type + '\n'
            indent += 2
          },
          leave: () => {
            indent -= 2
          }
        }
      })

      expect(visitedLog).toMatchSnapshot()
    }

    {
      let visitedLog = ''
      let indent = 0
      traverse(threeTwoThreeAst, {
        '*': {
          enter: (node, parent) => {
            visitedLog += ' '.repeat(indent) + node.type + '\n'
            indent += 2
          },
          leave: () => {
            indent -= 2
          }
        }
      })

      expect(visitedLog).toMatchSnapshot()
    }

    {
      let visitedLog = ''
      let indent = 0
      traverse(propertiesMultiAst, {
        '*': {
          enter: (node, parent) => {
            visitedLog += ' '.repeat(indent) + node.type + '\n'
            indent += 2
          },
          leave: () => {
            indent -= 2
          }
        }
      })

      expect(visitedLog).toMatchSnapshot()
    }
  })

  it('invokes all matching visitors for each node type', () => {
    {
      const visitedNodes: string[] = []
      traverse(basicAst, {
        EdgeChain: node => visitedNodes.push(node.type),
        Statement: node => visitedNodes.push(node.type),
        '*': node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }

    {
      const visitedNodes: string[] = []
      traverse(threeTwoThreeAst, {
        EdgeChain: node => visitedNodes.push(node.type),
        Statement: node => visitedNodes.push(node.type),
        '*': node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }

    {
      const visitedNodes: string[] = []
      traverse(propertiesMultiAst, {
        EdgeChain: node => visitedNodes.push(node.type),
        Statement: node => visitedNodes.push(node.type),
        '*': node => visitedNodes.push(node.type)
      })

      expect(visitedNodes).toMatchSnapshot()
    }
  })

  it('allows mutation of nodes', () => {
    const ast = _.cloneDeep(basicAst)
    traverse(ast, {
      Program: node => {
        ;(node as AST).meta['hello'] = 'Hello, world!'
      }
    })

    expect(ast.meta['hello']).toBe('Hello, world!')
  })

  it('throws if it encounters an unknown node type', () => {
    expect(() => {
      traverse({ type: 'foo' } as any, {
        '*': () => {}
      })
    }).toThrow(/unknown/i)
  })
})
