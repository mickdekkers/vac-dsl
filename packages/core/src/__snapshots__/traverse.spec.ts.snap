// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverse accepts aliases 1`] = `
Array [
  "EdgeChain",
]
`;

exports[`traverse accepts aliases 2`] = `
Array [
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "EdgeChain",
]
`;

exports[`traverse accepts aliases 3`] = `
Array [
  "VariableDeclaration",
  "VariableDeclaration",
  "VariableDeclaration",
  "EdgeChain",
  "EdgeChain",
]
`;

exports[`traverse invokes all matching visitors for each node type 1`] = `
Array [
  "Program",
  "EdgeChain",
  "EdgeChain",
  "EdgeChain",
  "NodeList",
  "Literal",
  "NodeList",
  "Literal",
]
`;

exports[`traverse invokes all matching visitors for each node type 2`] = `
Array [
  "Program",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "EdgeChain",
  "EdgeChain",
  "EdgeChain",
  "NodeList",
  "Identifier",
  "Identifier",
  "Identifier",
  "NodeList",
  "Identifier",
  "Identifier",
  "NodeList",
  "Identifier",
  "Identifier",
  "Identifier",
]
`;

exports[`traverse invokes all matching visitors for each node type 3`] = `
Array [
  "Program",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "VariableDeclaration",
  "VariableDeclaration",
  "Identifier",
  "Literal",
  "EdgeChain",
  "EdgeChain",
  "EdgeChain",
  "NodeList",
  "Identifier",
  "NodeList",
  "Identifier",
  "PropertyList",
  "Property",
  "EdgeChain",
  "EdgeChain",
  "EdgeChain",
  "NodeList",
  "Identifier",
  "NodeList",
  "Identifier",
  "PropertyList",
  "Property",
]
`;

exports[`traverse visits all nodes in order when using * catch-all 1`] = `
"Program
  EdgeChain
    NodeList
      Literal
    NodeList
      Literal
"
`;

exports[`traverse visits all nodes in order when using * catch-all 2`] = `
"Program
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  EdgeChain
    NodeList
      Identifier
      Identifier
      Identifier
    NodeList
      Identifier
      Identifier
    NodeList
      Identifier
      Identifier
      Identifier
"
`;

exports[`traverse visits all nodes in order when using * catch-all 3`] = `
"Program
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  VariableDeclaration
    Identifier
    Literal
  EdgeChain
    NodeList
      Identifier
    NodeList
      Identifier
    PropertyList
      Property
  EdgeChain
    NodeList
      Identifier
    NodeList
      Identifier
    PropertyList
      Property
"
`;
